File: python/src/aoc/day2.py
import re

PATTERN = r"^(\d+)-(\d+) (\w+):\s(.+)\s*$"


def _parse(line):
    return re.match(PATTERN, line).groups()


def _find_valid_passwords(data, validator):
    return [x for x in data if validator(*_parse(x))]


def _validator_1(min, max, char, password):
    return int(min) <= password.count(char) <= int(max)


def _get_characters(indices, string):
    return [string[int(i)-1] for i in indices]


def _validator_2(idx1, idx2, char, password):
    candidates = _get_characters([idx1, idx2], password)
    return sum([1 for c in candidates if c == char]) == 1


def solve(data):
    print('Part 1:', len(_find_valid_passwords(data, _validator_1)))
    print('Part 2:', len(_find_valid_passwords(data, _validator_2)))


def solve_1(data):
    return len(_find_valid_passwords(data, _validator_1))


def solve_2(data):
    return len(_find_valid_passwords(data, _validator_2))

File: clojure/src/aoc/day2.clj
(ns aoc.day2
  (:require [clojure.test :refer [deftest testing is run-tests]]
            [clojure.string :refer [split-lines]]))

(def example-data ["1-3 a: abcde"
                   "1-3 b: cdefg"
                   "2-9 c: ccccccccc"])
(def day2-data (split-lines (slurp "data/day2.data")))
(def pattern #"^(\d+)-(\d+) (\w+):\s(.+)\s*$")

(defn- parse-line [line]
  (let [[_ number-1 number-2 character password] (re-matches pattern line)]
    [(Integer/parseInt number-1)
     (Integer/parseInt number-2)
     (.charAt character 0)
     password]))

(defn- validate-part1 [[min max character password]]
  (let [count (get (frequencies password) character 0)]
    (and (>= count min)
         (<= count max))))

(defn- nth-chars [indexes string]
  (map #(nth string (dec %)) indexes))

(defn- validate-part2 [[pos-1 pos-2 character password]]
  (->> password
       (nth-chars [pos-1 pos-2])
       (filter (partial = character))
       count
       (= 1)))
    

(defn solve
  "Day2.1: Find number of password(s) compliant with their policy"
  [data validator]
  (->> data
       (map parse-line)
       (filter validator)
       count))

(deftest day-2-verification
  (testing "Part One"
    (is (= (solve example-data validate-part1) 2))
    (is (= (solve day2-data validate-part1) 569)))
  (testing "Part Two"
    (is (= (solve example-data validate-part2) 1))
    (is (= (solve day2-data validate-part2) 346))))

(run-tests)